crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -22,8 +22,13 @@ pub struct EntityRef<'w> {
 }
 
 impl<'w> EntityRef<'w> {
+    /// # Safety
+    ///
+    /// Entity and location _must_ be valid for the provided world.
     #[inline]
-    pub(crate) fn new(world: &'w World, entity: Entity, location: EntityLocation) -> Self {
+    pub(crate) unsafe fn new(world: &'w World, entity: Entity, location: EntityLocation) -> Self {
+        debug_assert!(world.entities().contains(entity));
```

Comment:
```suggestion
        debug_assert!(world.entities().contains(entity));
        debug_assert_eq!(world.entities().get(entity.index()), Some(location));
```

If we're checking that these are valid, we should also ensure the location is valid too.